eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1327 @@
+};
+
+typedef struct _EknrRendererPrivate
+{
+  GHashTable *cache;
+} EknrRendererPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (EknrRenderer,
+                            eknr_renderer,
+                            G_TYPE_OBJECT)
+
+enum {
+  PROP_0,
+  NPROPS
+};
+
+static GParamSpec *eknr_renderer_props [NPROPS] = { NULL, };
+
+typedef struct _RendererMustacheData {
+  GVariantDict  *variables;
+  GError       **error; /* non-owned */
+
+  mustache_str_ctx input;
+  mustache_str_ctx output;
+
+  const char *section_variable;
+} RendererMustacheData;
+
+static RendererMustacheData *
+renderer_mustache_data_new (GVariantDict  *variables,
+                            GError       **error,
+                            const char    *input)
+{
+  RendererMustacheData *data = g_new0 (RendererMustacheData, 1);
+  data->variables = variables ? g_variant_dict_ref (variables) : NULL;
+  data->error = error;
+  data->input.string = input ? g_strdup (input) : NULL;
+  data->input.offset = 0;
+  data->output.string = NULL;
+  data->output.offset = 0;
+
+  return data;
+}
+
+static void
+renderer_mustache_data_free (RendererMustacheData *data)
+{
+  g_clear_pointer (&data->variables, g_variant_dict_unref);
+  g_clear_pointer (&data->output.string, free);
+  g_clear_pointer (&data->input.string, g_free);
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (RendererMustacheData,
+                               renderer_mustache_data_free);
+
+static void
+free_mustache_template (mustache_template_t *template)
+{
+  mustache_api_t api = {
+    .freedata = NULL
+  };
+
+  mustache_free (&api, template);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (mustache_template_t, free_mustache_template)
+
+static uintmax_t
+_renderer_read_from_closure (mustache_api_t *api,
+                             void           *userdata,
+                             char           *buffer,
+                             uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strread (api, (void *) &data->input, buffer, buffer_size);
+}
+
+static uintmax_t
+_renderer_write_to_closure (mustache_api_t *api,
+                            void           *userdata,
+                            const char     *buffer,
+                            uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strwrite (api, (void *) &data->output, (char *) buffer, buffer_size);
+}
+
+static char *
+_lookup_in_gvariant_dict (RendererMustacheData *data,
+                          const char           *text,
+                          mustache_api_t       *api)
+{
+  g_autoptr(GVariant) value_v = g_variant_dict_lookup_value (data->variables, text, G_VARIANT_TYPE ("s"));
+  const char *value = NULL;
+
+  if (value_v == NULL)
+    {
+      g_autofree char *msg = g_strdup_printf ("No such variable %s", text);
+      (*api->error) (api, data, __LINE__, msg);
+      return NULL;
+    }
+
+  return g_variant_dup_string (value_v, NULL);
```

Comment:
This function does not need to return a copy of the string, and the caller should be changed accordingly.

Reply:
It does not indeed - thanks for spotting, saves a copy